### PR TITLE
Deprecate 'genRange', add tests

### PR DESCRIPTION
**Pull request by curiousleo**
Adds a default implementation for `next` and deprecates it.

This seems like a sensible way forward to me:
- RNG users can still call `next`, but are nudged towards `genWordX[R]`
- Doesn't break compatibility with current RNG implementations which implement `next`
- Allows RNG implementations to drop the `next` implementation

The idea being that in some future version, `next` is simply removed.

Alternatively, we could remove `next` altogether. Thoughts?

### CODE REVIEW

File: System/Random.hs [hunk 1]
```diff
@@ -131,8 +133,22 @@ class RandomGen g where
   -- |The 'next' operation returns an 'Int' that is uniformly distributed
   -- in the range returned by 'genRange' (including both end points),
   -- and a new generator.
-  next     :: g -> (Int, g)
-  -- `next` can be deprecated over time
+  {-# DEPRECATE next "Use genWord32[R] or genWord64[R]" #-}
+  next :: g -> (Int, g)
+  next g = (minR + fromIntegral w, g') where
+    (minR, maxR) = genRange g
+    range = fromIntegral $ maxR - minR
```
Comment (lehins) [inline System/Random.hs:144]: This implementation will overflow

Comment (lehins) [reply System/Random.hs:144]: Example:
```haskell
Prelude> maxR = maxBound ::Int
Prelude> maxR - (-1)
-9223372036854775808
```

Comment (lehins) [reply System/Random.hs:144]: Although due modulo arithmetics it might work out in the end

Comment (curiousleo) [reply System/Random.hs:144]: The type of `range` is `Word64` or `Word32`, depending on the size of `Int`.

Without a conversion, I also get an overflow:

```haskell
*System.Random> (maxBound :: Int) - (minBound :: Int)
-1
```

But the relevant case is where the type of `range` is known to be the word of the correct size:

```haskell
*System.Random> fromIntegral ((maxBound :: Int) - (minBound :: Int)) :: Word64
18446744073709551615
```

AFAICT it's the latter that happens on line 140.

**Comment by curiousleo** [comment]
Ready for another round of reviews.

**Comment by curiousleo** [comment]
Closing in favour of https://github.com/idontgetoutmuch/random/pull/35.

**Pull request closed without merging**